iroh-sync/src/sync.rs

```diff
@@ -1903,14 +1916,214 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn test_replica_queries_mem() -> Result<()> {
+        let store = store::memory::Store::default();
+
+        test_replica_queries(store)?;
+        Ok(())
+    }
+
+    #[cfg(feature = "fs-store")]
+    #[test]
+    fn test_replica_queries_fs() -> Result<()> {
+        let dbfile = tempfile::NamedTempFile::new()?;
+        let store = store::fs::Store::new(dbfile.path())?;
+        test_replica_queries(store)?;
+
+        Ok(())
+    }
+
+    fn test_replica_queries<S: store::Store>(store: S) -> Result<()> {
+        let mut rng = rand_chacha::ChaCha12Rng::seed_from_u64(1);
+        let namespace = Namespace::new(&mut rng);
+        let mut replica = store.new_replica(namespace)?;
+        let namespace = replica.namespace();
+
+        let a1 = store.new_author(&mut rng)?;
+        let a2 = store.new_author(&mut rng)?;
+        let a3 = store.new_author(&mut rng)?;
+        println!(
+            "a1 {} a2 {} a3 {}",
+            a1.id().fmt_short(),
+            a2.id().fmt_short(),
+            a3.id().fmt_short()
+        );
+
+        replica.hash_and_insert("hi/world", &a2, "a2")?;
+        replica.hash_and_insert("hi/world", &a1, "a1")?;
+        replica.hash_and_insert("hi/moon", &a2, "a1")?;
+        replica.hash_and_insert("hi", &a3, "a3")?;
+
+        let expect = |rows: &[(&str, &Author)]| {
+            rows.into_iter()
+                .map(|(key, author)| (key.to_string(), author.id()))
+                .collect::<Vec<_>>()
+        };
+
+        assert_eq!(
```

Comment:
could write the below a little more readable like this

```rust

struct Test<'a> {
  query: Query,
  res: Vec<(&'static str, &'a AuthorId)>,
}

let tests = [
  Test {  
     query: Query::single_latest_per_key(), 
     res: vec![("hi", &a3), ("hi/moon", &a2), ("hi/world", &a1)],
  },
  // ... 
];

for test in tests {
  assert_eq!(
    query(&store, namespace, test.query),
    expect(&test.res),
    "query: {:#?}", test.query
  );
}
```

Reply:
Came up with something similar independently when writing more tests a minute ago ;) but I think I like your's even better, will switch

Reply:
:)

Reply:
did something inbetween ;)